File: Graph/BFS.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

public class BFS {
    private ArrayList<ArrayList<Integer>> graph;
    private boolean[] visited;
    private ArrayList<Integer> traversalOrder;

    public BFS(int vertices) {
        graph = new ArrayList<>();
        for (int i = 0; i < vertices; i++) {
            graph.add(new ArrayList<>());
        }
    }

    // 양방향 간선 추가
    public void addEdge(int u, int v) {
        graph.get(u).add(v);
        graph.get(v).add(u);
    }

    // 단방향 간선 추가
    public void addDirectedEdge(int from, int to) {
        graph.get(from).add(to);
    }

    // 모든 vertex에 대한 BFS
    public ArrayList<Integer> traverseAll() {
        visited = new boolean[graph.size()];
        traversalOrder = new ArrayList<>();

        for (int vertex = 0; vertex < graph.size(); vertex++) {
            if (!visited[vertex]) {
                bfsIterative(vertex);
            }
        }

        return traversalOrder;
    }

    // 특정 vertex부터 BFS
    public ArrayList<Integer> traverseFrom(int startVertex) {
        visited = new boolean[graph.size()];
        traversalOrder = new ArrayList<>();
        bfsIterative(startVertex);
        return traversalOrder;
    }

    private void bfsIterative(int startVertex) {
        Queue<Integer> queue = new LinkedList<>();
        visited[startVertex] = true;
        queue.offer(startVertex);

        while (!queue.isEmpty()) {
            int currentVertex = queue.poll();
            traversalOrder.add(currentVertex);

            for (int neighbor : graph.get(currentVertex)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.offer(neighbor);
                }
            }
        }
    }
}

File: Graph/BFS.kt
import java.util.LinkedList
import java.util.Queue

class BFS(vertices: Int) {
    private val graph: MutableList<MutableList<Int>> = MutableList(vertices) { mutableListOf() }
    private lateinit var visited: BooleanArray
    private lateinit var traversalOrder: MutableList<Int>

    // 양방향 간선 추가
    fun addEdge(u: Int, v: Int) {
        graph[u].add(v)
        graph[v].add(u)
    }

    // 단방향 간선 추가
    fun addDirectedEdge(from: Int, to: Int) {
        graph[from].add(to)
    }

    // 모든 vertex에 대한 DFS
    fun traverseAll(): List<Int> {
        visited = BooleanArray(graph.size)
        traversalOrder = mutableListOf()

        for (vertex in graph.indices) {
            if (!visited[vertex]) {
                bfsIterative(vertex)
            }
        }

        return traversalOrder
    }

    // 특정 vertex부터 DFS
    fun traverseFrom(startVertex: Int): List<Int> {
        visited = BooleanArray(graph.size)
        traversalOrder = mutableListOf()
        bfsIterative(startVertex)
        return traversalOrder
    }

    private fun bfsIterative(startVertex: Int) {
        val queue: Queue<Int> = LinkedList()
        visited[startVertex] = true
        queue.offer(startVertex)

        while (queue.isNotEmpty()) {
            val currentVertex = queue.poll()
            traversalOrder.add(currentVertex)

            for (neighbor in graph[currentVertex]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true
                    queue.offer(neighbor)
                }
            }
        }
    }
}
